### PR TITLE
Print full paths in spellchecking suggestions when needed for disambiguation

### PR DESCRIPTION
**Pull request by nojb**
See [MPR#7864](https://caml.inria.fr/mantis/view.php?id=7864). It's a bit hackish, so suggestions welcome!

Before:
image

After:
image

### CODE REVIEW

**Comment by gasche** [comment]
I'm a bit disturbed by the proposed behavior: at the point in the program where the hint is given, `foo` and `A.foo` mean the same thing, right? (In fact `foo` refers to "the previous `foo` binding" but the user cannot access this.)

When I saw the bugreport, my idea for a fix was to simply accumulate a set of names (removing duplicates in the process), but I had no time to look at this.

**Comment by alainfrisch** [comment]
> foo and A.foo mean the same thing, right?

Not necessarily:

````ocaml
let foo = 0
module X = struct let foo = 1 end
module Y = struct let foo = 2 end
open X
open Y
let _ = fooo
`````

gives:
````
Hint: Did you mean foo, foo or foo?
````

Still, I wouldn't be shocked to suggest only `foo` (once).

**Comment by nojb** [comment]
I agree with @gasche. Initially I had been playing with an example as in @alainfrisch's note above, which is why this felt logical to me; but on second thought I agree it does not seem the right thing to do in general. I will implement the simpler alternative of simply eliminating duplicates.

**Comment by gasche** [comment]
( @alainfrisch I didn't immediately understand your point, but you mean that the proposed implementation would say `foo, X.foo or Y.foo`, and that only two out of those three paths go to the same definition at the point of error.)

File: utils/misc.ml [hunk 1]
```diff
@@ -546,7 +546,8 @@ let spellcheck env name =
          else if dist = best_dist then (head :: best_choice, dist)
          else acc
   in
-  fst (List.fold_left (compare name) ([], max_int) env)
+  fst (List.fold_left (compare name) ([], max_int)
+         (List.sort_uniq String.compare env))
```
Comment (Armael) [inline utils/misc.ml]: Currently, the order of the elements in the output list is the reverse of the order in the `env` input list. After this patch, it is set to the reverse of the alphabetical order (as can be seen in the currently failing test).

I think it's fine for this PR to change the order (the `env` input list is coming from a fold on the environment and hence in an unspecified order anyway). I would just slightly prefer if it was the alphabetical order, not the reversed alphabetical order :-).

Comment (nojb) [reply utils/misc.ml]: Thanks, fixed.

**Comment by Armael** [comment]
Could you maybe add a test demonstrating the fix? The code that you show in the initial message, for example.

**Review by Armael — APPROVED** [review]
In any case, I'm approving the PR. Feel free to merge if you feel like adding the test is not worth the effort :-).

**Comment by nojb** [comment]
Thanks! I will add a test, but haven't found the time yet.

**Comment by nojb** [comment]
Added a test. I didn't know where to put it (I didn't find any test dedicated to the spellchecking suggestion), so it ended up in `typing-core-bugs`...

**Comment by Armael** [comment]
Thanks!

**Pull request merged**